restic list pack snapshotID - create lists of packfiles for a given snapshot as text

wplapper opened the pull request:
Command to show packfiles which belong to a snapshot or list of snapshots.

packfilelist gathers all snapshots from the the standard restic.SnapshotFilter, collects the trees which belong to these snapshots and calls restic.FindUsedBlobs to collect all blobs participating in these snapshots. These blobs get translated to their containing packfiles and these lists are printed out either as standard text of in JSON format. The `--detail` options allows for more insights to be looked at.

<!--
Thank you very much for contributing code or documentation to restic! Please
fill out the following questions to make it easier for us to review your
changes.
-->

What does this PR change? What problem does it solve?
-----------------------------------------------------
This PR addresses issues raised in #5372.
<!--
Describe the changes and their purpose here, as detailed as needed.
-->

Was the change previously discussed in an issue or on the forum?
----------------------------------------------------------------

<!--
Link issues and relevant forum posts here.

If this PR resolves an issue on GitHub, use "Closes #1234" so that the issue
is closed automatically when this PR is merged.
-->
https://github.com/restic/restic/issues/5372
https://github.com/restic/restic/issues/5751

Closes #5372

Checklist
---------

<!--
You do not need to check all the boxes below all at once. Feel free to take
your time and add more commits. If you're done and ready for review, please
check the last box. Enable a checkbox by replacing [ ] with [x].

Please always follow these steps:
- Read the [contribution guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches).
- Enable [maintainer edits](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- Run `gofmt` on the code in all commits.
- Format all commit messages in the same style as [the other commits in the repository](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits).
-->

- [X] I have added tests for all code changes.
- [ ] I have added documentation for relevant changes (in the manual).
- [X] There's a new file in `changelog/unreleased/` that describes the changes for our users (see [template](https://github.com/restic/restic/blob/master/changelog/TEMPLATE)).
- [ ] I'm done! This pull request is ready for review.

wplapper commented:
I have decided to revert this PR to closed status to ease the pressure on the 
maintainers of restic. Once the number of PRs has dropped considerably, I might
moved this PR back to the queue to be reviewed. In case you consider this PR
important, please contact me via the usual channels. Thank you!

MichaelEischer reviewed (changes requested):
I seriously dislike adding an extra command for something that is conceptually a part of the `restore` command. Clarify the use case and sketch how a cleanly integrated solution could look like before making any further code changes.

The remarks below highlight additional problems, in addition to the overall design issue.

Please only add options that where it's clear what their use case is. Exposing arbitrary internal details of the repository format is a bad idea as this will hinder future evolution of the repository format.

As the original use case for this command was to get a list of `packfiles` that must be warmed up prior to a `restore` operation, it should focus on that. With that limited focus it is probably a much better fit as an extension to an existing command. Btw, the use case of getting a list of all pack files containing tree blobs should be solved via a `restic list treepacks` command.

Some further comments on the options:
- `--short-id`: I don't see where this would ever be useful. The command is highly technical so there's nothing to gain from short ids.
- `--summary`: Not useful for the mentioned use cases. Some of the information _might_ make sense for the `stats` command, but definitely not here.
- `--detail`: just pick one level and don't provide the others at all
- `--detail 2`: I don't even remotely see a use case for that. It also exposes too many internals.
- `--orphan`: Too many internal details
- `--no-header`: Nope. If that's a problem for someone then they just have to use the JSON output.

wplapper committed:
restric restore --packfiles

new implementation to produce a list of packfiles for a snapshot or
the whole repository

cmd/restic/cmd_restore.go:
interface to runPackfileList() added, option `--packfiles` added

cmd/restic/packfiles_from_snapshots.go:
reworked implementation to generate the packfile list

cmd/restic/cmd_restore_integration_test.go:
added tests for text and JSON output, plus helper functions to achieve it

doc/045_working_with_repos.rst:
added paragraph about creating a packfile list for a snapshot / repository

changelog/unreleased/issue-5372:
announcement of new enhancement

restore - restore original state of repository,

remove any changes made by my PR so far

restore - remove doc updates

removing the changes to the documentation

wplapper commented:
There is a need for such type of command / sub-command. See https://github.com/restic/restic/issues/5751.

MichaelEischer commented on cmd/restic/cmd_restore.go:
```diff
@@ -1,16 +1,26 @@
+	"crypto/aes"
+	"encoding/binary"
+	"encoding/json"
+	"golang.org/x/crypto/poly1305"
```
No code outside `./internal/repository` may interact with this kind of repository internals (not up for discussion). After https://github.com/restic/restic/pull/21848 this is also most likely not possible anymore anyways. If it were _absolutely_ necessary, we could discuss exposing `pack.Size` (which performs the same calculations as you do below). But I don't see that this is the case.

wplapper replied:
code removed.

MichaelEischer commented on cmd/restic/cmd_restore.go:
```diff
@@ -315,3 +353,159 @@ func getXattrSelectFilter(opts RestoreOptions, printer progress.Printer) (func(x
+}
+
+// produceJSONOutput generates JSON output
+func produceJSONOutput(packfilesSort []restic.ID, snapPacks map[restic.ID]PacklistInfo, stdout io.Writer) error {
```
Drop all output functionality beyond printing a plain list/array of pack file names. For the given use case of knowing which pack files need an Object Lock, it's neither important to learn the pack file type or size. The current approach to calculate the pack file size is also not compatible with my plans to encapsulate the restic internal anyways.

wplapper replied:
no JSON output anymore.

MichaelEischer reviewed (changes requested):
Sorry for the late review. It took me a while to figure out what I don't like about the current approach and how to address it. The main problem is that we have two use cases all mixed up:
- apply S3 object lock to all packs of some/all snapshots
- warm up files on cold storage

At first glance the warm-up use case just differs by only handling one snapshot at a time. But the difference is more fundamental: when only restoring part of a snapshot it is useful to only warm up that specific part. There's no equivalent for that in the S3 object lock case; one always has to protect the whole snapshot not just a subset.

The current PR is a better fit of the S3 object lock case, but that use case does not belong into the `restore` command. So different from what I've written initially, this code belongs somewhere else. We already have the `restic list packs` command. My feeling is that we should extend it to optionally accept a list of snapshots as parameters. That command only prints a plain text list of pack IDs, so there's also no need for extra information like type or size and also no need for a JSON output format.

As a result the `restic list packs <snapshots>` command will be usable to lock pack files of a snapshot and as a crutch to warm up the data packs for snapshot. Fine-grained warm ups for S3 already already implemented separately.

wplapper committed:
list packs <snapshotID> - reworked code

reworked code to generate a list of packfile IDs which are used by
a snapshot

cmd/restic/cmd_list.go:
integrated code as part of `restic list packs`

cmd/restic/cmd_list_integration_test.go:
adopted tests accordingly

wplapper committed:
new interface repo.LookupBlob()

wplapper committed:
list packs <snapshotID> -changelog

modified changelof:
command ist now attached tp `restic list packs`

wplapper committed:
list packs ... - documentation

cmd/restic/cmd_list.go: removed call to MemorizeLister, will be called
implicitly

added docu for `restic list packs snapshotID`

MichaelEischer committed:
cleanup parameter check and error handling

MichaelEischer committed:
polish code and changelog

MichaelEischer committed:
simplify packfile tests

MichaelEischer reviewed (approved):
LGTM. I've cleaned up the code a bit and simplified the test skaffolding. The check for the number of parameters passed to `list` is now also stricter again. Passing too many parameters now also returns an error.
